nixos/tests/pleroma.nix

```diff
@@ -0,0 +1,265 @@
+/*
+  Pleroma E2E VM test.
+
+  Abstract:
+  =========
+  Using pleroma, postgresql, a local CA cert, a nginx reverse proxy
+  and a toot-based client, we're going to:
+
+  1. Provision a pleroma service from scratch (pleroma config + postgres db).
+  2. Create a "jamy" admin user.
+  3. Send a toot from this user.
+  4. Send a upload from this user.
+  5. Check the toot is part of the server public timeline
+
+  Notes:
+  - We need a fully functional TLS setup without having any access to
+    the internet. We do that by issuing a self-signed cert, add this
+    self-cert to the hosts pki trust store and finally spoof the
+    hostnames using /etc/hosts.
+  - For this NixOS test, we *had* to store some DB-related and
+    pleroma-related secrets to the store. Keep in mind the store is
+    world-readable, it's the worst place possible to store *any*
+    secret. **DO NOT DO THIS IN A REAL WORLD DEPLOYMENT**.
+*/
+
+import ./make-test-python.nix ({ pkgs, ... }:
+  let
+  send-toot = pkgs.writeScriptBin "send-toot" ''
+    set -eux
+    # toot is using the requests library internally. This library
+    # sadly embed its own certificate store instead of relying on the
+    # system one. Overriding this pretty bad default behaviour.
+    export REQUESTS_CA_BUNDLE=/etc/ssl/certs/ca-certificates.crt
```

Comment:
Is this using its own bundled version of it, or the nixpkgs-provided version?
I wonder what happened to 768c970c8d339f76947976c31ad2c8723247a8f3, I'd assume requests was patched to use `
/etc/ssl/certs/ca-certificates.crt` instead…

Reply:
That's weird, it do not seem like `requests` gets overriden in the proccess: https://github.com/NixOS/nixpkgs/blob/54985aeaf714505611f83589e57236010baee408/pkgs/applications/misc/toot/default.nix#L17

However, if I remove this line, the test end up failing with:

```
client # Traceback (most recent call last):
client #   File "/nix/store/wsj5danxnka3zgiz155w1y02p4zax1n5-python3.8-urllib3-1.25.10/lib/python3.8/site-packages/urllib3/connectionpool.py", line 670, in urlopen
client #     httplib_response = self._make_request(
client #   File "/nix/store/wsj5danxnka3zgiz155w1y02p4zax1n5-python3.8-urllib3-1.25.10/lib/python3.8/site-packages/urllib3/connectionpool.py", line 381, in _make_request
client #     self._validate_conn(conn)
client #   File "/nix/store/wsj5danxnka3zgiz155w1y02p4zax1n5-python3.8-urllib3-1.25.10/lib/python3.8/site-packages/urllib3/connectionpool.py", line 978, in _validate_conn
client #     conn.connect()
client #   File "/nix/store/wsj5danxnka3zgiz155w1y02p4zax1n5-python3.8-urllib3-1.25.10/lib/python3.8/site-packages/urllib3/connection.py", line 362, in connect
client #     self.sock = ssl_wrap_socket(
client #   File "/nix/store/wsj5danxnka3zgiz155w1y02p4zax1n5-python3.8-urllib3-1.25.10/lib/python3.8/site-packages/urllib3/util/ssl_.py", line 384, in ssl_wrap_socket
client #     return context.wrap_socket(sock, server_hostname=server_hostname)
client #   File "/nix/store/gvkhgh813chlsvdsnxgwhv72xns08xcp-python3-3.8.6/lib/python3.8/ssl.py", line 500, in wrap_socket
client #     return self.sslsocket_class._create(
client #   File "/nix/store/gvkhgh813chlsvdsnxgwhv72xns08xcp-python3-3.8.6/lib/python3.8/ssl.py", line 1040, in _create
client #     self.do_handshake()
client #   File "/nix/store/gvkhgh813chlsvdsnxgwhv72xns08xcp-python3-3.8.6/lib/python3.8/ssl.py", line 1309, in do_handshake
client #     self._sslobj.do_handshake()
client # ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate (_ssl.c:1124)
client # 
client # During handling of the above exception, another exception occurred:
client # 
client # Traceback (most recent call last):
client #   File "/nix/store/nplv8hd6mljj3rymv7s6269q13y1j9gk-python3.8-requests-2.24.0/lib/python3.8/site-packages/requests/adapters.py", line 439, in send
client #     resp = conn.urlopen(
client #   File "/nix/store/wsj5danxnka3zgiz155w1y02p4zax1n5-python3.8-urllib3-1.25.10/lib/python3.8/site-packages/urllib3/connectionpool.py", line 726, in urlopen
client #     retries = retries.increment(
client #   File "/nix/store/wsj5danxnka3zgiz155w1y02p4zax1n5-python3.8-urllib3-1.25.10/lib/python3.8/site-packages/urllib3/util/retry.py", line 439, in increment
client #     raise MaxRetryError(_pool, url, error or ResponseError(cause))
client # urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='pleroma.nixos.test', port=443): Max retries exceeded with url: /api/v1/instance (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate (_ssl.c:1124)')))
client # 
client # During handling of the above exception, another exception occurred:
client # 
client # Traceback (most recent call last):
client #   File "/nix/store/hbwji1fddg9ndpaxm95nzwcbx2m0rm16-toot-0.27.0/bin/.toot-wrapped", line 9, in <module>
client #     sys.exit(main())
client #   File "/nix/store/hbwji1fddg9ndpaxm95nzwcbx2m0rm16-toot-0.27.0/lib/python3.8/site-packages/toot/console.py", line 550, in main
client #     run_command(app, user, command_name, args)
client #   File "/nix/store/hbwji1fddg9ndpaxm95nzwcbx2m0rm16-toot-0.27.0/lib/python3.8/site-packages/toot/console.py", line 532, in run_command
client #     return fn(app, user, parsed_args)
client #   File "/nix/store/hbwji1fddg9ndpaxm95nzwcbx2m0rm16-toot-0.27.0/lib/python3.8/site-packages/toot/commands.py", line 188, in login_cli
client #     app = create_app_interactive(instance=args.instance, scheme=args.scheme)
client #   File "/nix/store/hbwji1fddg9ndpaxm95nzwcbx2m0rm16-toot-0.27.0/lib/python3.8/site-packages/toot/auth.py", line 43, in create_app_interactive
client #     return config.load_app(instance) or register_app(instance, scheme)
client #   File "/nix/store/hbwji1fddg9ndpaxm95nzwcbx2m0rm16-toot-0.27.0/lib/python3.8/site-packages/toot/auth.py", line 15, in register_app
client #     instance = api.get_instance(domain, scheme)
client #   File "/nix/store/hbwji1fddg9ndpaxm95nzwcbx2m0rm16-toot-0.27.0/lib/python3.8/site-packages/toot/api.py", line 281, in get_instance
client #     return http.anon_get(url).json()
client #   File "/nix/store/hbwji1fddg9ndpaxm95nzwcbx2m0rm16-toot-0.27.0/lib/python3.8/site-packages/toot/http.py", line 62, in anon_get
client #     response = send_request(request)
client #   File "/nix/store/hbwji1fddg9ndpaxm95nzwcbx2m0rm16-toot-0.27.0/lib/python3.8/site-packages/toot/http.py", line 17, in send_request
client #     response = session.send(prepared, allow_redirects=allow_redirects, **settings)
client #   File "/nix/store/nplv8hd6mljj3rymv7s6269q13y1j9gk-python3.8-requests-2.24.0/lib/python3.8/site-packages/requests/sessions.py", line 643, in send
client #     r = adapter.send(request, **kwargs)
client #   File "/nix/store/nplv8hd6mljj3rymv7s6269q13y1j9gk-python3.8-requests-2.24.0/lib/python3.8/site-packages/requests/adapters.py", line 514, in send
client #     raise SSLError(e, request=request)
client # requests.exceptions.SSLError: HTTPSConnectionPool(host='pleroma.nixos.test', port=443): Max retries exceeded with url: /api/v1/instance (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate (_ssl.c:1124)')))
client: output: Creating config file at /root/.config/toot/config.json
Looking up instance info...

error: 
Traceback (most recent call last):
  File "/nix/store/nc5cfr15lavxwa8i334f1xnwsd08h019-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 899, in run_tests
    exec(tests, globals())
  File "<string>", line 1, in <module>
  File "<string>", line 7, in <module>
  File "/nix/store/nc5cfr15lavxwa8i334f1xnwsd08h019-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 422, in succeed
    raise Exception(
Exception: command `send-toot` failed (exit code 1)
cleaning up
killing client (pid 57)
killing pleroma (pid 9)
(0.00 seconds)
```

TL;DR: it's not reading my `/etc/ssl/certs` file.

I'll leave that line for now.

Reply:
Mystery solved: the patch is no more.

https://github.com/NixOS/nixpkgs/blob/54985aeaf714505611f83589e57236010baee408/pkgs/development/python-modules/requests/default.nix#L13

Reply:
Good point.